Another attempt at expanding attribution button

sargunv opened the pull request:
Based on #293, but drops the `Popup` in favor of just animating a `Surface`. Seems to avoid all the bugs discussed in that thread, and works the same on iOS and Android

Tested on iOS and Android with single line and multi line attributions with the button placed in all four corners. Also tested style switching and RTL layouts.

westnordost commented:
Well, that was the initial attempt, i.e. #248. If you think this is the better approach, I think it would make more sense to base an implementation on that PR, than #293. However, why did I pivot to #293? Because:

---

> Hm, I am a bit discontent with this. It is weird to have a component in the map HUD layout that can expand its size to any size (and thus displace other elements)
> 
> I'll look into whether/how to display the expanded attribution info in a view that overlaps all (popup / dialog), hopefully possible with a nice animation like it is now.

https://github.com/maplibre/maplibre-compose/pull/248#issuecomment-2731214808

The issue I had when trying out the #248 implementation was when the attribution thingy was not alone: In my layout, I have a couple of buttons and I want 

1. the **attribution button** to **align with other buttons** (e.g. in a column)
2. while the **attribution popup** should display **on top** of all that, i.e. not displace other elements in the layout

The only way this could be achieved was to place some placeholder the size of the attribution button into the layout that has the other button and then have another layout on top that would align with that. **Very hacky**.  So, I figured, a "fake" popup will create more problems in the long run than a real popup.

While, with #293 we have a chain of mostly minor issues:
1. using a standard popup would display *next to* the ℹ️-button. I changed that so that it displays on top and put another ℹ️ icon in the popup that aligns with the button to create the illusion that it is the same element. 
2. letting compose calculate the popup position itself led to the ℹ️ icon not aligning perfectly with the ℹ️ of the button. @michalgwo fixed that by providing a custom calculation
3. that custom calculation apparently caused some other issue on edge cases which is fixed in #346

sargunv commented:
Oh my bad, I should've looked at the first attempt before working on this one.

If it's in a column, the displacement only happens if the attribution is multiple lines (taller than the button alone), right? I think all the custom logic like SuperimposingPopupPositionProvider and the custom calculation for alignment isn't worth maintaining here for a probably niche case, especially as it's a straightforward workaround to place the attribution button in a `Box` and the column underneath with a spacer (much less complex than all the custom calculations in #293)

So I think we should provide either:
* something like this, an expanding element
* or, a popup with the standard popup position provider (display next to the button)

sargunv commented:
(but also, as a user, I'd prefer the UI to rearrange itself around an oversized attribution info, especially if the attribution info is open by default)

michalgwo commented:
When I rotate the screen on iOS, app crashes. Works fine on Android.
```
Uncaught Kotlin exception: androidx.compose.runtime.ComposeRuntimeError: Compose Runtime internal error. Unexpected or incorrect use of the Compose internal runtime API (pending composition has not been applied). Please report to Google or use https://goo.gle/compose-feedback
    at 0   DemoApp                             0x104bde19f        kfun:kotlin.Throwable#<init>(){} + 75 
    at 1   DemoApp                             0x104bd7cd3        kfun:kotlin.Exception#<init>(){} + 67 
    at 2   DemoApp                             0x104bd7ea3        kfun:kotlin.RuntimeException#<init>(){} + 67 
    at 3   DemoApp                             0x104bd8367        kfun:kotlin.IllegalStateException#<init>(){} + 67 
    at 4   DemoApp                             0x104e5c85f        kfun:androidx.compose.runtime.ComposeRuntimeError#<init>(kotlin.String){} + 87 
    at 5   DemoApp                             0x104e5cac3        kfun:androidx.compose.runtime#composeRuntimeError(kotlin.String){}kotlin.Nothing + 455 
    at 6   DemoApp                             0x104e62dff        kfun:androidx.compose.runtime.CompositionImpl.drainPendingModificationsForCompositionLocked#internal + 431 
    at 7   DemoApp                             0x104e63b13        kfun:androidx.compose.runtime.CompositionImpl#composeContent(kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){} + 835 
    at 8   DemoApp                             0x104fabc97        kfun:androidx.compose.runtime.ControlledComposition#composeContent(kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){}-trampoline + 99 
    at 9   DemoApp                             0x104e93f0f        kfun:androidx.compose.runtime.Recomposer#composeInitial(androidx.compose.runtime.ControlledComposition;kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){} + 911 
    at 10  DemoApp                             0x104faadf3        kfun:androidx.compose.runtime.CompositionContext#composeInitial(androidx.compose.runtime.ControlledComposition;kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){}-trampoline + 67 
    at 11  DemoApp                             0x104e549df        kfun:androidx.compose.runtime.ComposerImpl.CompositionContextImpl#composeInitial(androidx.compose.runtime.ControlledComposition;kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){} + 195 
    at 12  DemoApp                             0x104faadf3        kfun:androidx.compose.runtime.CompositionContext#composeInitial(androidx.compose.runtime.ControlledComposition;kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){}-trampoline + 67 
    at 13  DemoApp                             0x104e62c2b        kfun:androidx.compose.runtime.CompositionImpl.composeInitial#internal + 319 
    at 14  DemoApp                             0x104e62247        kfun:androidx.compose.runtime.CompositionImpl#setContent(kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){} + 243 
    at 15  DemoApp                             0x104fabb13        kfun:androidx.compose.runtime.ReusableComposition#setContent(kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){}-trampoline + 99 
    at 16  DemoApp                             0x1058e50e3        kfun:androidx.compose.ui.layout.LayoutNodeSubcompositionsState.subcomposeInto#internal + 467 
    at 17  DemoApp                             0x1058e4de7        kfun:androidx.compose.ui.layout.LayoutNodeSubcompositionsState.subcompose#internal.29 + 1151 
    at 18  DemoApp                             0x1058e4927        kfun:androidx.compose.ui.layout.LayoutNodeSubcompositionsState.subcompose#internal + 1107 
    at 19  DemoApp                             0x1058e43af        kfun:androidx.compose.ui.layout.LayoutNodeSubcompositionsState#subcompose(kotlin.Any?;kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){}kotlin.collections.List<androidx.compose.ui.layout.Measurable> + 2351 
    at 20  DemoApp                             0x1058e9d1b        kfun:androidx.compose.ui.layout.LayoutNodeSubcompositionsState.Scope.subcompose#internal + 135 
    at 21  DemoApp                             0x105b35c67        kfun:androidx.compose.ui.layout.SubcomposeMeasureScope#subcompose(kotlin.Any?;kotlin.Function2<androidx.compose.runtime.Composer,kotlin.Int,kotlin.Unit>){}kotlin.collections.List<androidx.compose.ui.layout.Measurable>-trampoline + 115 
    at 22  DemoApp                             0x105eda957        kfun:androidx.compose.material3.ScaffoldLayout$$inlined$cache$1.invoke#internal + 9567 
    at 23  DemoApp                             0x105edcd23        kfun:androidx.compose.material3.ScaffoldLayout$$inlined$cache$1.$<bridge-NNNNU>invoke(androidx.compose.ui.layout.SubcomposeMeasureScope;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult#internal + 163 
    at 24  DemoApp                             0x104d0d5a3        kfun:kotlin.Function2#invoke(1:0;1:1){}1:2-trampoline + 115 
    at 25  DemoApp                             0x1058ec68b        kfun:androidx.compose.ui.layout.LayoutNodeSubcompositionsState.LayoutNodeSubcompositionsState$createMeasurePolicy$1.measure#internal + 1163 
    at 26  DemoApp                             0x105b338ff        kfun:androidx.compose.ui.layout.MeasurePolicy#measure__at__androidx.compose.ui.layout.MeasureScope(kotlin.collections.List<androidx.compose.ui.layout.Measurable>;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 27  DemoApp                             0x10590751b        kfun:androidx.compose.ui.node.InnerNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 1179 
    at 28  DemoApp                             0x105b3bca3        kfun:androidx.compose.ui.node.NodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 67 
    at 29  DemoApp                             0x105959113        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.invoke#internal + 183 
    at 30  DemoApp                             0x10595917f        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.$<bridge-DNN>invoke(){}#internal + 71 
    at 31  DemoApp                             0x104d0a83f        kfun:kotlin.Function0#invoke(){}1:0-trampoline + 99 
    at 32  DemoApp                             0x104f5ae73        kfun:androidx.compose.runtime.snapshots.Snapshot.Companion#observe(kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function0<0:0>){0§<kotlin.Any?>}0:0 + 1111 
    at 33  DemoApp                             0x104f97c1b        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver.ObservedScopeMap.observe#internal + 891 
    at 34  DemoApp                             0x104f94b93        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<kotlin.Any>} + 1383 
    at 35  DemoApp                             0x105988c9b        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<androidx.compose.ui.node.OwnerScope>} + 227 
    at 36  DemoApp                             0x1059889b3        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeMeasureSnapshotReads(androidx.compose.ui.node.LayoutNode;kotlin.Boolean;kotlin.Function0<kotlin.Unit>){} + 411 
    at 37  DemoApp                             0x1059547cf        kfun:androidx.compose.ui.node.MeasurePassDelegate#performMeasure(androidx.compose.ui.unit.Constraints){} + 571 
    at 38  DemoApp                             0x105954f9b        kfun:androidx.compose.ui.node.MeasurePassDelegate#remeasure(androidx.compose.ui.unit.Constraints){}kotlin.Boolean + 995 
    at 39  DemoApp                             0x105954b6f        kfun:androidx.compose.ui.node.MeasurePassDelegate#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 743 
    at 40  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 41  DemoApp                             0x105bc54fb        kfun:androidx.compose.foundation.layout.BoxMeasurePolicy.measure#internal + 1063 
    at 42  DemoApp                             0x105b338ff        kfun:androidx.compose.ui.layout.MeasurePolicy#measure__at__androidx.compose.ui.layout.MeasureScope(kotlin.collections.List<androidx.compose.ui.layout.Measurable>;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 43  DemoApp                             0x10590751b        kfun:androidx.compose.ui.node.InnerNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 1179 
    at 44  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 45  DemoApp                             0x10587f32b        kfun:androidx.compose.ui.graphics.SimpleGraphicsLayerModifier.measure#internal + 219 
    at 46  DemoApp                             0x105b386e7        kfun:androidx.compose.ui.node.LayoutModifierNode#measure__at__androidx.compose.ui.layout.MeasureScope(androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 47  DemoApp                             0x10590f38f        kfun:androidx.compose.ui.node.LayoutModifierNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 2491 
    at 48  DemoApp                             0x105b3bca3        kfun:androidx.compose.ui.node.NodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 67 
    at 49  DemoApp                             0x105959113        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.invoke#internal + 183 
    at 50  DemoApp                             0x10595917f        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.$<bridge-DNN>invoke(){}#internal + 71 
    at 51  DemoApp                             0x104d0a83f        kfun:kotlin.Function0#invoke(){}1:0-trampoline + 99 
    at 52  DemoApp                             0x104f5ae73        kfun:androidx.compose.runtime.snapshots.Snapshot.Companion#observe(kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function0<0:0>){0§<kotlin.Any?>}0:0 + 1111 
    at 53  DemoApp                             0x104f97c1b        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver.ObservedScopeMap.observe#internal + 891 
    at 54  DemoApp                             0x104f94b93        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<kotlin.Any>} + 1383 
    at 55  DemoApp                             0x105988c9b        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<androidx.compose.ui.node.OwnerScope>} + 227 
    at 56  DemoApp                             0x1059889b3        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeMeasureSnapshotReads(androidx.compose.ui.node.LayoutNode;kotlin.Boolean;kotlin.Function0<kotlin.Unit>){} + 411 
    at 57  DemoApp                             0x1059547cf        kfun:androidx.compose.ui.node.MeasurePassDelegate#performMeasure(androidx.compose.ui.unit.Constraints){} + 571 
    at 58  DemoApp                             0x105954f9b        kfun:androidx.compose.ui.node.MeasurePassDelegate#remeasure(androidx.compose.ui.unit.Constraints){}kotlin.Boolean + 995 
    at 59  DemoApp                             0x105954b6f        kfun:androidx.compose.ui.node.MeasurePassDelegate#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 743 
    at 60  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 61  DemoApp                             0x105c3c483        kfun:androidx.compose.animation.AnimatedEnterExitMeasurePolicy.measure#internal + 867 
    at 62  DemoApp                             0x105b338ff        kfun:androidx.compose.ui.layout.MeasurePolicy#measure__at__androidx.compose.ui.layout.MeasureScope(kotlin.collections.List<androidx.compose.ui.layout.Measurable>;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 63  DemoApp                             0x10590751b        kfun:androidx.compose.ui.node.InnerNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 1179 
    at 64  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 65  DemoApp                             0x105c51adb        kfun:androidx.compose.animation.EnterExitTransitionModifierNode.measure#internal + 1783 
    at 66  DemoApp                             0x105b386e7        kfun:androidx.compose.ui.node.LayoutModifierNode#measure__at__androidx.compose.ui.layout.MeasureScope(androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 67  DemoApp                             0x10590f38f        kfun:androidx.compose.ui.node.LayoutModifierNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 2491 
    at 68  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 69  DemoApp                             0x10587daff        kfun:androidx.compose.ui.graphics.BlockGraphicsLayerModifier#measure__at__androidx.compose.ui.layout.MeasureScope(androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult + 219 
    at 70  DemoApp                             0x105b386e7        kfun:androidx.compose.ui.node.LayoutModifierNode#measure__at__androidx.compose.ui.layout.MeasureScope(androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 71  DemoApp                             0x10590f38f        kfun:androidx.compose.ui.node.LayoutModifierNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 2491 
    at 72  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 73  DemoApp                             0x105c33573        kfun:androidx.compose.animation.AnimatedContent$$inlined$fastForEach$1.AnimatedContent$$inlined$fastForEach$1$invoke$1.invoke#internal + 219 
    at 74  DemoApp                             0x105c3390b        kfun:androidx.compose.animation.AnimatedContent$$inlined$fastForEach$1.AnimatedContent$$inlined$fastForEach$1$invoke$1.$<bridge-NNNNNU>invoke(androidx.compose.ui.layout.MeasureScope;androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult#internal + 199 
    at 75  DemoApp                             0x104d0d697        kfun:kotlin.Function3#invoke(1:0;1:1;1:2){}1:3-trampoline + 123 
    at 76  DemoApp                             0x1058d17e7        kfun:androidx.compose.ui.layout.LayoutModifierImpl#measure__at__androidx.compose.ui.layout.MeasureScope(androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult + 267 
    at 77  DemoApp                             0x105b386e7        kfun:androidx.compose.ui.node.LayoutModifierNode#measure__at__androidx.compose.ui.layout.MeasureScope(androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 78  DemoApp                             0x10590f38f        kfun:androidx.compose.ui.node.LayoutModifierNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 2491 
    at 79  DemoApp                             0x105b3bca3        kfun:androidx.compose.ui.node.NodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 67 
    at 80  DemoApp                             0x105959113        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.invoke#internal + 183 
    at 81  DemoApp                             0x10595917f        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.$<bridge-DNN>invoke(){}#internal + 71 
    at 82  DemoApp                             0x104d0a83f        kfun:kotlin.Function0#invoke(){}1:0-trampoline + 99 
    at 83  DemoApp                             0x104f5ae73        kfun:androidx.compose.runtime.snapshots.Snapshot.Companion#observe(kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function0<0:0>){0§<kotlin.Any?>}0:0 + 1111 
    at 84  DemoApp                             0x104f97c1b        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver.ObservedScopeMap.observe#internal + 891 
    at 85  DemoApp                             0x104f94b93        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<kotlin.Any>} + 1383 
    at 86  DemoApp                             0x105988c9b        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<androidx.compose.ui.node.OwnerScope>} + 227 
    at 87  DemoApp                             0x1059889b3        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeMeasureSnapshotReads(androidx.compose.ui.node.LayoutNode;kotlin.Boolean;kotlin.Function0<kotlin.Unit>){} + 411 
    at 88  DemoApp                             0x1059547cf        kfun:androidx.compose.ui.node.MeasurePassDelegate#performMeasure(androidx.compose.ui.unit.Constraints){} + 571 
    at 89  DemoApp                             0x105954f9b        kfun:androidx.compose.ui.node.MeasurePassDelegate#remeasure(androidx.compose.ui.unit.Constraints){}kotlin.Boolean + 995 
    at 90  DemoApp                             0x105954b6f        kfun:androidx.compose.ui.node.MeasurePassDelegate#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 743 
    at 91  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 92  DemoApp                             0x105c2ed5b        kfun:androidx.compose.animation.AnimatedContentMeasurePolicy.measure#internal + 1715 
    at 93  DemoApp                             0x105b338ff        kfun:androidx.compose.ui.layout.MeasurePolicy#measure__at__androidx.compose.ui.layout.MeasureScope(kotlin.collections.List<androidx.compose.ui.layout.Measurable>;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 94  DemoApp                             0x10590751b        kfun:androidx.compose.ui.node.InnerNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 1179 
    at 95  DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 96  DemoApp                             0x105c278df        kfun:androidx.compose.animation.AnimatedContentTransitionScopeImpl.SizeModifierNode.measure#internal + 347 
    at 97  DemoApp                             0x105b386e7        kfun:androidx.compose.ui.node.LayoutModifierNode#measure__at__androidx.compose.ui.layout.MeasureScope(androidx.compose.ui.layout.Measurable;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 98  DemoApp                             0x10590f38f        kfun:androidx.compose.ui.node.LayoutModifierNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 2491 
    at 99  DemoApp                             0x105b3bca3        kfun:androidx.compose.ui.node.NodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 67 
    at 100 DemoApp                             0x105959113        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.invoke#internal + 183 
    at 101 DemoApp                             0x10595917f        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.$<bridge-DNN>invoke(){}#internal + 71 
    at 102 DemoApp                             0x104d0a83f        kfun:kotlin.Function0#invoke(){}1:0-trampoline + 99 
    at 103 DemoApp                             0x104f5ae73        kfun:androidx.compose.runtime.snapshots.Snapshot.Companion#observe(kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function0<0:0>){0§<kotlin.Any?>}0:0 + 1111 
    at 104 DemoApp                             0x104f97c1b        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver.ObservedScopeMap.observe#internal + 891 
    at 105 DemoApp                             0x104f94b93        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<kotlin.Any>} + 1383 
    at 106 DemoApp                             0x105988c9b        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<androidx.compose.ui.node.OwnerScope>} + 227 
    at 107 DemoApp                             0x1059889b3        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeMeasureSnapshotReads(androidx.compose.ui.node.LayoutNode;kotlin.Boolean;kotlin.Function0<kotlin.Unit>){} + 411 
    at 108 DemoApp                             0x1059547cf        kfun:androidx.compose.ui.node.MeasurePassDelegate#performMeasure(androidx.compose.ui.unit.Constraints){} + 571 
    at 109 DemoApp                             0x105954f9b        kfun:androidx.compose.ui.node.MeasurePassDelegate#remeasure(androidx.compose.ui.unit.Constraints){}kotlin.Boolean + 995 
    at 110 DemoApp                             0x105954b6f        kfun:androidx.compose.ui.node.MeasurePassDelegate#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 743 
    at 111 DemoApp                             0x105b2614b        kfun:androidx.compose.ui.layout.Measurable#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 107 
    at 112 DemoApp                             0x1059e3543        kfun:androidx.compose.ui.layout.OverlayLayout$1.measure#internal + 999 
    at 113 DemoApp                             0x105b338ff        kfun:androidx.compose.ui.layout.MeasurePolicy#measure__at__androidx.compose.ui.layout.MeasureScope(kotlin.collections.List<androidx.compose.ui.layout.Measurable>;androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.MeasureResult-trampoline + 123 
    at 114 DemoApp                             0x10590751b        kfun:androidx.compose.ui.node.InnerNodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable + 1179 
    at 115 DemoApp                             0x105b3bca3        kfun:androidx.compose.ui.node.NodeCoordinator#measure(androidx.compose.ui.unit.Constraints){}androidx.compose.ui.layout.Placeable-trampoline + 67 
    at 116 DemoApp                             0x105959113        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.invoke#internal + 183 
    at 117 DemoApp                             0x10595917f        kfun:androidx.compose.ui.node.MeasurePassDelegate.MeasurePassDelegate$1.$<bridge-DNN>invoke(){}#internal + 71 
    at 118 DemoApp                             0x104d0a83f        kfun:kotlin.Function0#invoke(){}1:0-trampoline + 99 
    at 119 DemoApp                             0x104f5ae73        kfun:androidx.compose.runtime.snapshots.Snapshot.Companion#observe(kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function1<kotlin.Any,kotlin.Unit>?;kotlin.Function0<0:0>){0§<kotlin.Any?>}0:0 + 1111 
    at 120 DemoApp                             0x104f97c1b        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver.ObservedScopeMap.observe#internal + 891 
    at 121 DemoApp                             0x104f94b93        kfun:androidx.compose.runtime.snapshots.SnapshotStateObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<kotlin.Any>} + 1383 
    at 122 DemoApp                             0x105988c9b        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeReads(0:0;kotlin.Function1<0:0,kotlin.Unit>;kotlin.Function0<kotlin.Unit>){0§<androidx.compose.ui.node.OwnerScope>} + 227 
    at 123 DemoApp                             0x1059889b3        kfun:androidx.compose.ui.node.OwnerSnapshotObserver#observeMeasureSnapshotReads(androidx.compose.ui.node.LayoutNode;kotlin.Boolean;kotlin.Function0<kotlin.Unit>){} + 411 
    at 124 DemoApp                             0x1059547cf        kfun:androidx.compose.ui.node.MeasurePassDelegate#performMeasure(androidx.compose.ui.unit.Constraints){} + 571 
    at 125 DemoApp                             0x105954f9b        kfun:androidx.compose.ui.node.MeasurePassDelegate#remeasure(androidx.compose.ui.unit.Constraints){}kotlin.Boolean + 995 
    Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [androidx.compose.ui.scene.MotionDurationScaleImpl@103c958, androidx.compose.runtime.BroadcastFrameClock@118c468, StandaloneCoroutine{Cancelling}@107dd10, FlushCoroutineDispatcher@114c708]
        at 0   DemoApp                             0x104bde027        kfun:kotlin.Throwable#<init>(kotlin.String?){} + 99 
        at 1   DemoApp                             0x104bd7d47        kfun:kotlin.Exception#<init>(kotlin.String?){} + 95 
        at 2   DemoApp                             0x104bd7f17        kfun:kotlin.RuntimeException#<init>(kotlin.String?){} + 95 
        at 3   DemoApp                             0x104e1ed63        kfun:kotlinx.coroutines.internal.DiagnosticCoroutineContextException#<init>(kotlin.coroutines.CoroutineContext){} + 175 
        at 4   DemoApp                             0x104df1233        kfun:kotlinx.coroutines.internal#handleUncaughtCoroutineException(kotlin.coroutines.CoroutineContext;kotlin.Throwable){} + 667 
        at 5   DemoApp                             0x104d8f57b        kfun:kotlinx.coroutines#handleCoroutineException(kotlin.coroutines.CoroutineContext;kotlin.Throwable){} + 735 
        at 6   DemoApp                             0x104d82e13        kfun:kotlinx.coroutines.StandaloneCoroutine.handleJobException#internal + 167 
        at 7   DemoApp                             0x104e264af        kfun:kotlinx.coroutines.JobSupport#handleJobException(kotlin.Throwable){}kotlin.Boolean-trampoline + 59 
        at 8   DemoApp                             0x104d9a6a7        kfun:kotlinx.coroutines.JobSupport.finalizeFinishingState#internal + 1227 
        at 9   DemoApp                             0x104da309f        kfun:kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath#internal + 2447 
        at 10  DemoApp                             0x104da26d7        kfun:kotlinx.coroutines.JobSupport.tryMakeCompleting#internal + 755 
        at 11  DemoApp                             0x104da21df        kfun:kotlinx.coroutines.JobSupport#makeCompletingOnce(kotlin.Any?){}kotlin.Any? + 451 
        at 12  DemoApp                             0x104d80c7f        kfun:kotlinx.coroutines.AbstractCoroutine#resumeWith(kotlin.Result<1:0>){} + 203 
        at 13  DemoApp                             0x104d0cf0f        kfun:kotlin.coroutines.Continuation#resumeWith(kotlin.Result<1:0>){}-trampoline + 99 
        at 14  DemoApp                             0x104be4267        kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 1295 
        at 15  DemoApp                             0x104d0cf0f        kfun:kotlin.coroutines.Continuation#resumeWith(kotlin.Result<1:0>){}-trampoline + 99 
        at 16  DemoApp                             0x104e1730f        kfun:kotlinx.coroutines.UndispatchedCoroutine#afterResume(kotlin.Any?){} + 243 
        at 17  DemoApp                             0x104e2283b        kfun:kotlinx.coroutines.AbstractCoroutine#afterResume(kotlin.Any?){}-trampoline + 59 
        at 18  DemoApp                             0x104d80cdb        kfun:kotlinx.coroutines.AbstractCoroutine#resumeWith(kotlin.Result<1:0>){} + 295 
        at 19  DemoApp                             0x104d0cf0f        kfun:kotlin.coroutines.Continuation#resumeWith(kotlin.Result<1:0>){}-trampoline + 99 
        at 20  DemoApp                             0x104be4267        kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 1295 
        at 21  DemoApp                             0x104d0cf0f        kfun:kotlin.coroutines.Continuation#resumeWith(kotlin.Result<1:0>){}-trampoline + 99 
        at 22  DemoApp                             0x104dfd5d3        kfun:kotlinx.coroutines.internal.ScopeCoroutine#afterResume(kotlin.Any?){} + 243 
        at 23  DemoApp                             0x104e2283b        kfun:kotlinx.coroutines.AbstractCoroutine#afterResume(kotlin.Any?){}-trampoline + 59 
        at 24  DemoApp                             0x104d80cdb        kfun:kotlinx.coroutines.AbstractCoroutine#resumeWith(kotlin.Result<1:0>){} + 295 
        at 25  DemoApp                             0x104d0cf0f        kfun:kotlin.coroutines.Continuation#resumeWith(kotlin.Result<1:0>){}-trampoline + 99 
        at 26  DemoApp                             0x104be4267        kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 1295 
        at 27  DemoApp                             0x104d0cf0f        kfun:kotlin.coroutines.Continuation#resumeWith(kotlin.Result<1:0>){}-trampoline + 99 
        at 28  DemoApp                             0x104df54cb        kfun:kotlinx.coroutines.DispatchedTask#run(){} + 1787 
        at 29  DemoApp                             0x104e257d3        kfun:kotlinx.coroutines.Runnable#run(){}-trampoline + 91 
        at 30  DemoApp                             0x1059fb203        kfun:androidx.compose.ui.platform.FlushCoroutineDispatcher.FlushCoroutineDispatcher$dispatch$1.FlushCoroutineDispatcher$dispatch$1$invoke$2.invoke#internal + 527 
        at 31  DemoApp                             0x1059fb2e3        kfun:androidx.compose.ui.platform.FlushCoroutineDispatcher.FlushCoroutineDispatcher$dispatch$1.FlushCoroutineDispatcher$dispatch$1$invoke$2.$<bridge-DNN>invoke(){}#internal + 71 
        at 32  DemoApp                             0x104d0a83f        kfun:kotlin.Function0#invoke(){}1:0-trampoline + 99 
        at 33  DemoApp                             0x1059fa7ff        kfun:androidx.compose.ui.platform.FlushCoroutineDispatcher.performRun#internal + 371 
        at 34  DemoApp                             0x1059faf0f        kfun:androidx.compose.ui.platform.FlushCoroutineDispatcher.FlushCoroutineDispatcher$dispatch$1.invoke#internal + 239 
        at 35  DemoApp                             0x104d0d5a3        kfun:kotlin.Function2#invoke(1:0;1:1){}1:2-trampoline + 115 
        at 36  DemoApp                             0x104be78f7        kfun:kotlin.coroutines.intrinsics.createCoroutineUnintercepted$$inlined$createCoroutineFromSuspendFunction$4.invokeSuspend#internal + 871 
        at 37  DemoApp                             0x104d0ce2f        kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#invokeSuspend(kotlin.Result<kotlin.Any?>){}kotlin.Any?-trampoline + 67 
        at 38  DemoApp                             0x104be400b        kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 691 
        at 39  DemoApp                             0x104d0cf0f        kfun:kotlin.coroutines.Continuation#resumeWith(kotlin.Result<1:0>){}-trampoline + 99 
        at 40  DemoApp                             0x104df55c7        kfun:kotlinx.coroutines.DispatchedTask#run(){} + 2039 
        at 41  DemoApp                             0x104e257d3        kfun:kotlinx.coroutines.Runnable#run(){}-trampoline + 91 
        at 42  DemoApp                             0x104e21457        kfun:kotlinx.coroutines.DarwinMainDispatcher.DarwinMainDispatcher$dispatch$$inlined$autoreleasepool$1.invoke#internal + 71 
        at 43  DemoApp                             0x104e214b3        kfun:kotlinx.coroutines.DarwinMainDispatcher.DarwinMainDispatcher$dispatch$$inlined$autoreleasepool$1.$<bridge-DNN>invoke(){}#internal + 71 
        at 44  DemoApp                             0x104d0a83f        kfun:kotlin.Function0#invoke(){}1:0-trampoline + 99 
        at 45  DemoApp                             0x104e225b7        _6f72672e6a6574627261696e732e6b6f746c696e783a6b6f746c696e782d636f726f7574696e65732d636f72652f6f70742f6275696c644167656e742f776f726b2f343465633665383530643563363366302f6b6f746c696e782d636f726f7574696e65732d636f72652f6e617469766544617277696e2f7372632f44697370617463686572732e6b74_knbridge13 + 191 
        at 46  libdispatch.dylib                   0x100d20583        _dispatch_call_block_and_release + 31 
        at 47  libdispatch.dylib                   0x100d3a063        _dispatch_client_callout + 15 
        at 48  libdispatch.dylib                   0x100d5af6f        _dispatch_main_queue_drain.cold.5 + 875 
        at 49  libdispatch.dylib                   0x100d30547        _dispatch_main_queue_drain + 179 
        at 50  libdispatch.dylib                   0x100d30483        _dispatch_main_queue_callback_4CF + 43 
        at 51  CoreFoundation                      0x18e36dd8f        <redacted> + 15 
        at 52  CoreFoundation                      0x18e3114f3        <redacted> + 1979 
        at 53  CoreFoundation                      0x18e312c3b        CFRunLoopRunSpecific + 571 
        at 54  GraphicsServices                    0x1db4f1453        GSEventRunModal + 167 
        at 55  UIKitCore                           0x190d25273        <redacted> + 815 
        at 56  UIKitCore                           0x190cf0a27        UIApplicationMain + 335 
        at 57  SwiftUI                             0x192e357a3        <redacted> + 167 
        at 58  SwiftUI                             0x192b3b01b        <redacted> + 111 
        at 59  SwiftUI                             0x192b3aecf        $s7SwiftUI3AppPAAE4mainyyFZ + 179 
        at 60  maplibre-compose-demo.debug.dylib   0x1001a89bb        $s21maplibre_compose_demo6iOSAppV5$mainyyFZ + 39 
        at 61  maplibre-compose-demo.debug.dylib   0x1001a8a67        __debug_main_executable_dylib_entry_point + 11 
        at 62  dyld                                0x1b51e7f07        <redacted> + 6039 
```

westnordost reviewed (commented):
Huh, I am puzzled how you made it so that the surface is not visible while not expanded, didn't notice anything to this effect when reading the code.

sargunv commented on lib/maplibre-compose-material3/src/commonMain/kotlin/dev/sargunv/maplibrecompose/material3/controls/AttributionButton.kt:
```diff
@@ -1,64 +1,208 @@
+  textLinkStyles: TextLinkStyles? = null,
+  shape: Shape = RoundedCornerShape(24.dp),
+  border: BorderStroke? = null,
+  expandedStyle: AttributionButtonStyle =
+    AttributionButtonStyle(
+      containerColor = MaterialTheme.colorScheme.surface,
+      contentColor = contentColorFor(MaterialTheme.colorScheme.surface),
+      tonalElevation = 0.dp,
+      shadowElevation = 0.dp,
+    ),
+  collapsedStyle: AttributionButtonStyle =
+    AttributionButtonStyle(
+      containerColor = MaterialTheme.colorScheme.surface.copy(alpha = 0f),
+      contentColor = contentColorFor(MaterialTheme.colorScheme.surface).copy(alpha = 0f),
+      tonalElevation = 0.dp,
+      shadowElevation = 0.dp,
+    ),
```
There's some composables used, at best it could be a composable function, and then we lose the "default value" display on mouse over in the IDE

sargunv commented:
> I am puzzled how you made it so that the surface is not visible while not expanded

The default collapsed color is transparent

sargunv commented:
> When I rotate the screen on iOS, app crashes.

Oof, definitely looks like a Compose bug. We should figure out a minimal repro to report

sargunv commented:
> When I wanted to place the AttributionButton in a Row, instead of breaking the line, it shrinks the other composable.

Hmm, I wonder if we can tweak the compass modifier to prefer breaking the FlowRow before shrinking the Button

sargunv commented:
Weirdly, the crash is reproducible with just this in the composable:

```
styleState.sources.flatMap { it.attributionLinks }
```

If I remove the call to `flatMap`, no crash.

It starts with: 

```
androidx.compose.runtime.ComposeRuntimeError: Compose Runtime internal error. Unexpected or incorrect use of the Compose internal runtime API (pending composition has not been applied). Please report to Google or use https://goo.gle/compose-feedback
```

and then, ten seconds layer, this (followed by the full stack trace posted previously):

```
Uncaught Kotlin exception: 2025-05-24 13:01:19.558816-0700 maplibre-compose-demo[13457:156020] void SendDelegateMessage(RetainPtr<NSInvocation> &&): delegate (webView:didReceiveViewportArguments:) failed to return after waiting 10.000000 seconds. main run loop mode: kCFRunLoopDefaultMode
2025-05-24 13:01:29.560403-0700 maplibre-compose-demo[13457:156020] void SendDelegateMessage(RetainPtr<NSInvocation> &&): delegate (webView:didReceiveViewportArguments:) failed to return after waiting 10.000000 seconds. main run loop mode: kCFRunLoopDefaultMode
```

sargunv commented:
Re: crash on rotation: I've worked around the crash in #357, though I think it's a bug of some sort in MapLibre Native for iOS, something to do with how it uses webkit to parse out the attribution html.

Re: the shrinking compass button, I guess the CompassButton should have `requiredSize` instead of `size`. Will make that change (though I think Compose will still move the button in that case). Still, I'd recommend using a Box layout with an `align` modifier for this kind of layout, so nothing shrinks:

![CleanShot 2025-05-24 at 17 11 16@2x](https://github.com/user-attachments/assets/cc4c58a4-431e-4a6c-9954-5b4fe7abecd0)

Re: column layout with Box hack, it's as simple as:

```kt
    Box(modifier = modifier.fillMaxSize().padding(8.dp)) {
      DisappearingScaleBar(
        metersPerDp = cameraState.metersPerDpAtTarget,
        zoom = cameraState.position.zoom,
        modifier = Modifier.align(Alignment.TopStart),
        measures = scaleBarMeasures,
      )
      Column(
        modifier = Modifier.align(Alignment.BottomEnd),
        verticalArrangement = Arrangement.spacedBy(8.dp),
        horizontalAlignment = Alignment.End,
      ) {
        CompassButton(cameraState, onClick = onCompassClick)
        Box(modifier = Modifier.requiredSize(48.dp)) // space for attribution button
      }
      AttributionButton(
        cameraState,
        styleState,
        modifier = Modifier.align(Alignment.BottomEnd),
        contentAlignment = Alignment.BottomEnd,
      )
    }
```

![CleanShot 2025-05-24 at 17 23 50@2x](https://github.com/user-attachments/assets/3daf9724-c6d6-4748-ab60-7bdca887f32f)

The only weird bit is the shadowless surface on top of the elevated button. While one can set a shadow with the current implementation, the animation is weird with transparency when collapsed (as is the default). If that's a problem, we can create a simpler `Popup` variant additionally.

sargunv commented:
I've refactored the component to make customization more flexible without having to add a whole bunch of parameters. The user can now swap in their own composables for the button, attribution, etc.

this also means it's a bit easier to reuse these partial components for alternative attribution styles. So I renamed this one to `ExpandingAttributionButton`; happy to accept a PopupAttributionButton or other styles in another PR

westnordost commented:
only if you set a fixed width for the flow row, otherwise it takes as much space as possible, IIRC

El 25 de mayo de 2025 12:05:34 CEST, "Michał Gwóźdź" ***@***.***> escribió:
>michalgwo left a comment (maplibre/maplibre-compose#349)
>
>> Re: the shrinking compass button, I guess the CompassButton should have `requiredSize` instead of `size`. Will make that change (though I think Compose will still move the button in that case). Still, I'd recommend using a Box layout with an `align` modifier for this kind of layout, so nothing shrinks:
>
>Is it an intended behavior? Shouldn't FlowRow break the line instead of pushing/shrinking other elements?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/maplibre/maplibre-compose/pull/349#issuecomment-2907733691
>You are receiving this because your review was requested.
>
>Message ID: ***@***.***>

michalgwo commented:
Looks like when the first attribution occupies 2 lines, the second attribution goes to to next line.

``` kotlin
  val attributions = listOf(AttributionLink("This is very very very very very very very very very long Attribution 1", ""), AttributionLink("Attribution 2", ""))
```

![image](https://github.com/user-attachments/assets/beddccbe-5a91-4d6e-9e81-35114be5d268)

sargunv commented:
Honestly there's surprisingly many edge cases with this that it's probably worth starting here with writing some comprehensive automatic tests, as I definitely got tired testing all the alignments, layout direction, and other interactions manually 😅

sargunv commented on lib/maplibre-compose-material3/src/commonMain/kotlin/dev/sargunv/maplibrecompose/material3/controls/AttributionButton.kt:
```diff
@@ -1,64 +1,237 @@
-      contentDescription = stringResource(Res.string.attribution),
-    )
-  }
+  val expanded = remember { MutableTransitionState(true) }
```
Perhaps an overload that takes expanded as a parameter instead of managing it internally? And then the caller can do whatever they like with it

michalgwo replied:
Yep

sargunv replied:
https://github.com/maplibre/maplibre-compose/pull/361
